### DISCUSSION TITLE
[Docker][SYCL] Making a minimal runtime Docker for SYCL for Intel graphics

### DISCUSSION THREAD

**Discussion by AnonymousRoboticist** [intel/llvm#6690] [Q&A]
I'm looking to make a multi-stage Docker build with the minimal amount of things in my image I need in order to run SYCL code at runtime on Intel graphics. Say if I based it on one of the sycl preinstalled dockers, I know I need to `COPY` over at least /opt/sycl which will have the compiler. I will need to apt install the intel opencl drivers (intel-opencl-icd). What about the /runtime folder that exists the current dockers? Is there any other thing I need to compiler code? 

It's all running on a physical device, so the image needs to be as minimal as possible. I currently have a simple dockerfile that copies over /opt/sycl/, adds to PATH and LD_LIBRARY_PATH, adds a sycl user with the right permissions (copied commands from the existing dockerfiles) 

```
# Original nightly image with everything installed
ARG base_tag=latest
ARG base_image=ghcr.io/intel/llvm/sycl_ubuntu2004_nightly

FROM $base_image:$base_tag

# Second stage
FROM ubuntu:20.04
RUN apt-get update && apt-get install build-essential -y
RUN apt-get install intel-opencl-icd -y

RUN mkdir -p /opt/sycl/
COPY --from=0 /opt/sycl/ /opt/sycl/

ENV DEBIAN_FRONTEND=noninteractive
ENV PATH /opt/sycl/bin:$PATH
ENV LD_LIBRARY_PATH /opt/sycl/lib:$LD_LIBRARY_PATH

RUN groupadd -g 1001 sycl && useradd sycl -u 1001 -g 1001 -m -s /bin/bash
RUN usermod -aG video sycl
RUN echo "sycl  ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers
COPY vector_add.cpp /vector_add.cpp
```
With this I am able to see my devices in sycl-ls, and clang++ with -fsycl compiles my code, but running it shows it crashes partway through (no exceptions caught). What else am I missing?

**Comment by AlexeySachkov** [comment]
Hi @AnonymousRoboticist,

Note: I'm might be answering your questions in arbitrary order, hope that this is not too confusing.

> but running it shows it crashes partway through (no exceptions caught). 

That is weird, do you have a backtrace of that crash? Launching an app under `SYCL_PI_TRACE=-1` could also help to pinpoint the problem.

> What else am I missing?

>  I will need to apt install the intel opencl drivers (intel-opencl-icd)

That is not a full driver and to be more precise, this is not a driver at all (even though it is called this way). ICD is a [loader](https://github.com/KhronosGroup/OpenCL-ICD-Loader) which performs OpenCL implementations lookup and dispatches calls from your app to selected implementation. It allows you to only link with a single `libOpenCL.so`, but use any number of OpenCL implementations from different vendors without re-linking.

To get an actual OpenCL or Level Zero implementation for Intel GPUs, you need to install packages from [intel/compute-runtime](https://github.com/intel/compute-runtime) repo. For example, here is the latest available release to this date: [`22.43.24558`](https://github.com/intel/compute-runtime/releases/tag/22.43.24558)

> Is there any other thing I need to compiler code?

In order to compile code, you don't need anything except for compiler, headers and `libsycl.so` to link with it.

> What about the /runtime folder that exists the current dockers? 

I'm not experienced with docker, but from what I see in our scripts:

`/runtime` is set to env variable `INSTALL_LOCATION`

https://github.com/intel/llvm/blob/3667cf3632eb7cf22656abf9019ed1d835d4c917/devops/containers/ubuntu2004_intel_drivers.Dockerfile#L21-L22

And if you look at [`install_drivers.sh`](https://github.com/intel/llvm/blob/sycl/devops/scripts/install_drivers.sh) it is only used to install TBB, OpenCL CPU RT and OpenCL FPGA Emulator. 

Graphics driver and runtimes are installed into some default location (presumably encoded in the packages):

https://github.com/intel/llvm/blob/3667cf3632eb7cf22656abf9019ed1d835d4c917/devops/scripts/install_drivers.sh#L37-L53

Here we separately install IGC (AOT and JIT compiler for Intel GPUs), NEO (compute runtime): OpenCL & Level Zero runtime and CM (that is another compiler which is used by ESIMD feature only, If I'm not mistaken).

From that list it should be enough to only install NEO (compute runtime), because their releases (and instructions) already include IGC, just follow links I posted above.
